presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -542,15 +632,61 @@ public Duration getTotalCpuTime()
     public void start()
     {
         if (started.compareAndSet(false, true)) {
-            executor.submit(this::schedule);
+            startScheduling(getStagesReadyForExecution());
+        }
+    }
+
+    private List<SqlStageExecution> getStagesReadyForExecution()
+    {
+        return stream(forTree(StreamingPlanSection::getChildren).depthFirstPreOrder(sectionedPlan))
+                .filter(this::isReadyForExecution)
+                .flatMap(section -> stream(forTree(SubPlan::getChildren).depthFirstPreOrder(section.getPlan())))
+                .map(SubPlan::getFragment)
+                .map(PlanFragment::getId)
+                .map(this::getFragmentExecution)
+                .collect(toImmutableList());
+    }
+
+    private boolean isReadyForExecution(StreamingPlanSection section)
+    {
+        SqlStageExecution stage = getFragmentExecution(section.getPlan().getFragment().getId());
+        if (stage.getState() != PLANNED) {
```

Comment:
Curious: Looks like the state will be `PLANNED` whenever `StageStateMachine` is created?  So is it to rule out already scheduled stages? If so add a comment like "already start"